src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,307 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  EMPTY,
+  merge as observableMerge,
+  Observable,
+  of as observableOf,
+  Subject,
+  Subscription,
+} from "rxjs";
+import {
+  catchError,
+  distinctUntilChanged,
+  filter,
+  finalize,
+  map,
+  mapTo,
+  mergeMap,
+  shareReplay,
+  switchMap,
+  take,
+  tap,
+} from "rxjs/operators";
+import { Representation } from "../../../manifest";
+import arrayFindIndex from "../../../utils/array_find_index";
+import {
+  areRangesOverlapping,
+  convertToRanges,
+} from "../../../utils/ranges";
+import request from "../../../utils/request";
+import log from "./log";
+import prepareSourceBuffer from "./media_source";
+
+interface IThumbnailInfo {
+  start: number;
+  duration: number;
+  mediaURL: string;
+}
+
+interface IThumbnailTrack {
+  thumbnailInfos: IThumbnailInfo[];
+  initURL: string;
+  codec: string;
+}
+
+/**
+ * Load needed segment data.
+ * @param {Object} thumbnails
+ * @param {HTMLMediaElement} mediaElement
+ */
+function getSegmentsData(
+  thumbnails: IThumbnailInfo[],
+  mediaElement: HTMLMediaElement
+): Observable<ArrayBuffer> {
+
+  const thumbnailsToLoad = thumbnails.filter((t) => {
+    const tRange = { start: t.start, end: t.start + t.duration };
+    const mediaRanges = convertToRanges(mediaElement.buffered);
+    return arrayFindIndex(mediaRanges, (mr) => {
+      return tRange.start >= mr.start && tRange.end <= mr.end;
+    }) === -1;
```

Comment:
Too difficult to read and not even optimal when compared to a more straightforward one.
Why not just iterating by hand ?